### PR TITLE
Division by zero error in the robust solver

### PR DESCRIPTION
**Pull request by ulricharmel**
I have made a few changes to avoid division by zero. I add a small error value to denominators to avoid division by zero. Also added few lines of codes to transfer flags to the robust gain machine so that these can be propagated to the robust weights.

### CODE REVIEW

**Comment by bennahugo** [comment]
@ulricharmel looks like there is an error in your stacking logic:
```
 - 22:05:53 - solver             [x14] [1.3/4.9 6.3/9.9 3.9Gb] Solver for tile 0 chunk D0T0F40 failed with exception: index 32768 is out of bounds for axis 1 with size 32
 - 22:05:53 - solver             [x14] [1.3/4.9 6.3/9.9 3.9Gb] Traceback (most recent call last):
  File "/scratch/bhugo/CubiCal/cubical/solver.py", line 793, in run_solver
    corr_vis, stats, outweights = solver(vdm, soldict, label, sol_opts)
  File "/scratch/bhugo/CubiCal/cubical/solver.py", line 565, in solve_and_correct
    _, stats, outweights = _solve_gains(vdm.gm, vdm.weighted_obser, vdm.weighted_model, vdm.flags_arr, sol_opts, label=label)
  File "/scratch/bhugo/CubiCal/cubical/solver.py", line 220, in _solve_gains
    gm.compute_update(model_arr, obser_arr)
  File "/scratch/bhugo/CubiCal/cubical/machines/complex_W_2x2_machine.py", line 170, in compute_update
    self.weights, self.v = self.update_weights(covinv, self.weights, self.v)
  File "/scratch/bhugo/CubiCal/cubical/machines/complex_W_2x2_machine.py", line 275, in update_weights
    self.weights[:,self.new_flags] = 0
IndexError: index 32768 is out of bounds for axis 1 with size 32
```

**Comment by ulricharmel** [comment]
@bennahugo can I see your full log file. I want to see the settings for your data tchunks. It seems to work on my simulations

**Comment by ulricharmel** [comment]
@bennahugo I think i have spotted the error. @o-smirnov please what is exactly in flags_arr. I thought it was kind of mask with only True (1) or False (0) values but it appears, it can be filled with any interger value and the flags will correspond to interger values different from zero

**Comment by bennahugo** [comment]
```
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [data] Visibility data options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - ms ................................................ = 1934_longtrack.ms
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - column ............................................ = DATA
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - time-chunk ........................................ = 32
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - freq-chunk ........................................ = 32
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - rebin-time ........................................ = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - rebin-freq ........................................ = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - chunk-by .......................................... = SCAN_NUMBER
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - chunk-by-jump ..................................... = 1.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - single-chunk ...................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [sel] Data selection options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - field ............................................. = 2
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - ddid .............................................. = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - taql .............................................. =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - chan .............................................. =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [model] Calibration model options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - list .............................................. = input/pks-1934-63-only.txt
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - ddes .............................................. = auto
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - beam-pattern ...................................... = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - beam-l-axis ....................................... = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - beam-m-axis ....................................... = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [montblanc] Montblanc simulation options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - device-type ....................................... = CPU
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - dtype ............................................. = float
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - feed-type ......................................... = linear
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - mem-budget ........................................ = 1024
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - verbosity ......................................... = WARNING
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - threads ........................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [weight] Weighting options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - column ............................................ = WEIGHT
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [flags] General flagging options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - apply ............................................. = -cubical
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - auto-init ......................................... = legacy
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - save .............................................. = cubical
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - save-legacy ....................................... = auto
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - reinit-bitflags ................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - warn-thr .......................................... = 0.3
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - see-no-evil ....................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [postmortem] Options for "postmortem" flagging based on solution statistics
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - enable ............................................ = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - tf-chisq-median ................................... = 1.2
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - tf-np-median ...................................... = 0.5
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - time-density ...................................... = 0.5
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - chan-density ...................................... = 0.5
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - ddid-density ...................................... = 0.5
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [madmax] Options for the "Mad Max" flagger
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - enable ............................................ = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - estimate .......................................... = corr
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - diag .............................................. = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - offdiag ........................................... = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - threshold ......................................... = [0, 20, 10]
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - global-threshold .................................. = [0, 25, 15]
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - plot .............................................. = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - plot-frac-above ................................... = 0.01
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - plot-bl ........................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - flag-ant .......................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - flag-ant-thr ...................................... = 5
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [sol] Solution options which apply at the solver level
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - jones ............................................. = G
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - precision ......................................... = 32
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - delta-g ........................................... = 1e-06
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - delta-chi ......................................... = 1e-06
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - chi-int ........................................... = 5
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - last-rites ........................................ = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - stall-quorum ...................................... = 0.99
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - diag-diag ......................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - term-iters ........................................ =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - min-bl ............................................ = 300.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - max-bl ............................................ = 0.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - subset ............................................ =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [bbc] Options for baseline-based corrections (a.k.a. BBCs, a.k.a. interferometer gains).
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - load-from ......................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - compute-2x2 ....................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - apply-2x2 ......................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - save-to ........................................... = {data[ms]}/BBC-field_{sel[field]}-ddid_{sel[ddid]}.parmdb
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - per-chan .......................................... = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - plot .............................................. = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [dist] Parallelization and distribution options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - ncpu .............................................. = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - nworker ........................................... = 32
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - nthread ........................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - max-chunks ........................................ = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - min-chunks ........................................ = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - pin ............................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - pin-io ............................................ = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - pin-main .......................................... = io
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [out] Options for output products
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - name .............................................. = pks1934
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - mode .............................................. = sc
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - column ............................................ = CORRECTED_DATA
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - model-column ...................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - weight-column ..................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - reinit-column ..................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - subtract-model .................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - subtract-dirs ..................................... = :
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - plots ............................................. = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - casa-gaintables ................................... = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [log] Options related to logging
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - memory ............................................ = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - boring ............................................ = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - append ............................................ = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - verbose ........................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - file-verbose ...................................... = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [debug] Debugging options for the discerning masochist
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - pdb ............................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - panic-amplitude ................................... = 0.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - stop-before-solver ................................ = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - escalate-warnings ................................. = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [misc] Miscellaneous options
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - random-seed ....................................... = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - parset-version .................................... = 0.1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [g] Options for G-Jones term
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - solvable .......................................... = True
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - type .............................................. = robust-2x2
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - load-from ......................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - xfer-from ......................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - save-to ........................................... = {data[ms]}/{JONES}-field_{sel[field]}-ddid_{sel[ddid]}.parmdb
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - dd-term ........................................... = 0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - fix-dirs .......................................... =
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - diag-diag ......................................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - update-type ....................................... = full
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - time-int .......................................... = 10
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - freq-int .......................................... = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - max-prior-error ................................... = 0.1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - max-post-error .................................... = 0.1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - clip-low .......................................... = 0.01
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - clip-high ......................................... = 10.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - clip-after ........................................ = 10
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - max-iter .......................................... = 1000
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - conv-quorum ....................................... = 0.99
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - ref-ant ........................................... = None
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - prop-flags ........................................ = default
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - robust-cov ........................................ = compute
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - robust-npol ....................................... = 2.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - robust-int ........................................ = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - robust-save-weights ............................... = False
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb] [de] de
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - dd-term ........................................... = 1
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - clip-low .......................................... = 0.0
 - 22:04:49 - main               [0.1/0.1 4.9/4.9 0.0Gb]  - clip-high ......................................... = 0.0
```

**Comment by ulricharmel** [comment]
I made a push check if it still fails @bennahugo

**Comment by o-smirnov** [comment]
> @o-smirnov please what is exactly in flags_arr. I thought it was kind of mask with only True (1) or False (0) values but it appears, it can be filled with any interger value and the flags will correspond to interger values different from zero

Apologies, @ulricharmel, only spotted this now. Your second guess is correct, flags are integers. Each buit represents a different flag condition. Anything with ``(flags_arr!=0)`` is to be considered flagged.

**Comment by bennahugo** [comment]
retest this please

**Comment by ulricharmel** [comment]
Hi @bennahugo are you still get the issue with  divide by zero or did I make a change that is conflicitng with the current code

**Comment by bennahugo** [comment]
@ulricharmel @o-smirnov ok I can confirm this branch is running. I think the casacore issue is only coming in if you're adding a new column (I didn't run into it when writing out CORRECTED_DATA, put in place by a previous CASA clearcal call). Should have some first results later today. Just one note the solver is now extremely verbose. I presume this will cleaned before the branch is merged, eg?
```
 - 21:42:43 - solver             [x04] [0.8/1.5 3.6/4.2 1.4Gb] D0T14F71 solver flags GOOB:71(46.41%) GVAR:10(6.54%) LOWSNR:1(0.65%): 0 (0.00%) new data flags
 - 21:42:43 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1460 iters: v-parameter is  2.0
 - 21:42:44 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1465 iters: v-parameter is  2.0
 - 21:42:44 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1470 iters: v-parameter is  2.0
 - 21:42:45 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1475 iters: v-parameter is  2.0
 - 21:42:46 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1480 iters: v-parameter is  2.0
 - 21:42:47 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1485 iters: v-parameter is  2.0
 - 21:42:48 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1490 iters: v-parameter is  2.0
 - 21:42:48 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1495 iters: v-parameter is  2.0
 - 21:42:49 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] dd : 1500 iters: v-parameter is  2.0
 - 21:42:49 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] D0T14F78 dd: 1500 iters, conv 77.78%, g/fl 52.29%, d/fl 27.45%, PGE 0.09 0.0 0.639, NFPGE 0 0 3, stall 62.38%, chi^2 1.485 -> 1.136 (1.135), noise 0.395 -> 0.395
 - 21:42:49 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] D0T14F78 solver flags GOOB:74(48.37%) GVAR:3(1.96%) ILLCOND:3(1.96%): 0 (0.00%) new data flags
 - 21:44:05 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 15 iters: v-parameter is  100.0
 - 21:46:32 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 20 iters: v-parameter is  100.0
 - 21:49:04 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 25 iters: v-parameter is  100.0
 - 21:51:28 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 30 iters: v-parameter is  100.0
 - 21:53:57 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 35 iters: v-parameter is  100.0
 - 21:56:28 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 40 iters: v-parameter is  100.0
 - 21:59:13 - solver             [x23] [0.4/1.5 3.1/4.2 3.0Gb] dd : 45 iters: v-parameter is  100.0
 - 22:01:45 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 50 iters: v-parameter is  100.0
 - 22:04:14 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 55 iters: v-parameter is  100.0
 - 22:06:46 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 60 iters: v-parameter is  100.0
 - 22:09:11 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 65 iters: v-parameter is  100.0
 - 22:11:40 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 70 iters: v-parameter is  100.0
 - 22:14:11 - solver             [x23] [0.4/1.5 3.1/4.2 1.4Gb] dd : 75 iters: v-parameter is  100.0

```

**Comment by o-smirnov** [comment]
Yeah it needs to log at a higher verbosity level. I notice this:

```
 - 21:42:49 - solver             [x12] [0.5/1.5 3.2/4.2 1.4Gb] D0T14F78 solver flags GOOB:74(48.37%) GVAR:3(1.96%) ILLCOND:3(1.96%): 0 (0.00%) new data flags
```

GOOB means out of bounds means you have high and low clips set on the DD term. This is not a good idea I think...

**Comment by bennahugo** [comment]
@o-smirnov it completely diverges without the clips because of residual RFI in some of the channels, hence the reason why I'm resorting to the robust solver

**Comment by o-smirnov** [comment]
And have you got madmax enabled? I'd like to look at the divergence a bit closer.

In any case disable the low clip, the DD term can go arbitrarily low (think beam null).

**Comment by bennahugo** [comment]
ok. No mad flags basically 90+% of the data

**Comment by bennahugo** [comment]
(at least the master version does)

**Comment by o-smirnov** [comment]
> ok. No mad flags basically 90+% of the data

Maybe madmax is the sane one in this scenario. :)

I'd like to debug this case deeply. Can you please give me the MS and a parset and a chunk ID where the divergence/90% flagging happens?

**Comment by bennahugo** [comment]
hmm you can have a tarball with 2GC or IFR corrected data. 
hugo@stills ~/CLUSTER_SURVEY/A3628.TriA.PSZ1G018.75+23.5                                                                                                                                                                                                             [9:51:06] 
> $ ls -lah reasonable_final2gc.tar.gz                                                                                                                                                                                                                            [±master ●●]
-rw-rw-r-- 1 hugo hugo 72G Mar 30 17:04 reasonable_final2gc.tar.gz
hugo@stills ~/CLUSTER_SURVEY/A3628.TriA.PSZ1G018.75+23.5                                                                                                                                                                                                             [9:51:12] 
> $ ls -lah reasonable_ifr.tar.gz                                                                                                                                                                                                                                 [±master ●●]
-rw-rw-r-- 1 hugo hugo 94G Mar 31 17:36 reasonable_ifr.tar.gz

This is before:
![image](https://user-images.githubusercontent.com/7934586/55461734-72083600-55f5-11e9-83e0-a551916bba1d.png)
and after:
![image](https://user-images.githubusercontent.com/7934586/55461754-80565200-55f5-11e9-9218-a3478ab937d1.png)
and before
![image](https://user-images.githubusercontent.com/7934586/55461868-d1664600-55f5-11e9-951c-e9998d44fc45.png)

and after
![image](https://user-images.githubusercontent.com/7934586/55461892-ddea9e80-55f5-11e9-929a-dbfd8bf3e6d4.png)

with the following parameters:
```

+def decalibrate(incol="SUBTRACTED_DATA", 
+              calincol="CORRECTED_DATA",
+              outcol="SUBTRACTED_DATA", 
+              model="MODEL_WITHOUT_DES",
+              lsmfilepostfix="decal1", 
+              des="{0:s}-catalog.lsm.html.de_tagged.lsm.html@dE",
+              label='decal',
+              freq_int=64,
+              masksig=[25, 25, 30],
+              solvemode='complex-diag',
+              corrtype='sr',
+              interval=[80, 80, 100],
+              restore=None):
+    for ti, t in enumerate(TARGET):
+        f = t + lsmfilepostfix
+        recipe.add("cab/cubical", "calibrate_target_%d" % ti, {
+            'data-ms': "%s.%s.1GC.ms" % (PREFIX, t),
+            'data-column': calincol,
+            'data-time-chunk': 100,
+            'data-freq-chunk': freq_int,
+            'model-list': [model,"{0:s}:output".format(des).format(f)],
+            'model-beam-pattern': "'MeerKAT_VBeam_10MHz_53Chans_$(corr)_$(reim).fits':output",
+            'model-beam-l-axis' : "X",
+            'model-beam-m-axis' : "Y",
+            'weight-column': "WEIGHT",
+            'flags-apply': "-cubical",
+            'flags-auto-init': "legacy",
+            'madmax-enable': False,
+            'sol-jones': 'dd',
+            'sol-min-bl': 110.0,
+            'sol-max-bl': 0,
+            'out-name': t,
+            'out-mode': corrtype,
+            'out-column': outcol,
+            'dd-time-int': interval[ti],
+            'dd-freq-int': freq_int,
+            'dd-type': solvemode,
+            'dd-max-iter': 1000,
+            'dist-ncpu': 16,
+            'dd-clip-high': 0,
+            'dd-clip-low': 0,
+#            'dd-fix-dirs': "0",
+            'out-subtract-dirs': '1:',
+            'dd-max-iter': 1000,
+            'dd-dd-term': True
+        }, input=INPUT, output=OUTPUT, label="calibrate %s %s" % (label, t), shared_memory="250g")

lsm and images are in the output folder in that same directory
```

**Comment by ulricharmel** [comment]
@bennahugo about the verbosity @o-smirnov wanted me to print the v -parameter of the robust solver. A low v-parameter implies outliers and high value no outliers. So may be that print line can be control with verbosity

**Comment by bennahugo** [comment]
Hmm maybe yes or print min, Q2, Q3, Q4 and max. But it is minor. I'm mostly seeing a slew of 2s.  As for the OOB I've changed the low clip to 0 and it converges faster ofc should have been obvious. Maybe that should have a fat warning printing. Anyways there are still the odd 20% going above 1.4 which I can only think is due to RFI, as amplitude time gains are already applied

**Comment by bennahugo** [comment]
btw it will be nice to know what all these acronyms stands for @o-smirnov... PGE... NFPGE....

**Comment by o-smirnov** [comment]
> @bennahugo about the verbosity @o-smirnov wanted me to print the v -parameter of the robust solver.

Yep. But do e.g. print>>log(2),"whatever", so that the printing is only enabled when you enable it with the appropriate --log-verbose setting.

**Comment by ulricharmel** [comment]
ok @o-smirnov cool. Will change that

**Comment by o-smirnov** [comment]
> btw it will be nice to know what all these acronyms stands for @o-smirnov... PGE... NFPGE....

MGE is the Maximum prior Gain Error, which is estimated from SNR considerations [here](https://github.com/ratt-ru/CubiCal/blob/master/cubical/machines/interval_gain_machine.py#L382) (and, I suspect, incorrectly....) This is one number per direction-correlation-solution interval. Intervals with prior GE > -max-prior-error (if the latter is set to non-0) are flagged to begin with (usiong the GSNR flag) and never solved for. The value printed for MGE is the max error across all intervals. NFMGE is the Number [of intervals] Flagged due to this.

PGE is the posterior gain error, estinated from the diagonal of Hessian. (Also incorrectly, I suspect...) This is now per direction-antenna-correlation-solution interval. Slots where PGE is in excess of -max-post-error are flagged (using the GVAR flag). The value printed is the max error across all antennas and intervals. NFPGE is the number of slots flagged.

Basically anything over .1 looks highly suspicious, especially for G-Jones, because then we're correcting data by the inverse of a highly uncertain value. For DD-Jones, higher errors are probably ok (especially when peeling weak soucres where some intervals will be very poorly conditioned) since we're only subtracting the best fit, and therefore conditioning is less of an issue.

**Comment by JSKenyon** [comment]
@bennahugo, can I close this? I believe these all went into master in the degridder PR.

**Comment by JSKenyon** [comment]
Files changed seems to be zero. So I am closing and killing this branch.
